endpoints/v2/__init__.py

```diff
@@ -128,44 +128,79 @@ def callback(results, response):
 
 
 def _require_repo_permission(permission_class, scopes=None, allow_public=False):
-    def wrapper(func):
-        @wraps(func)
-        def wrapped(namespace_name, repo_name, *args, **kwargs):
-            logger.debug(
-                "Checking permission %s for repo: %s/%s",
-                permission_class,
-                namespace_name,
-                repo_name,
-            )
-
-            permission = permission_class(namespace_name, repo_name)
-            if permission.can():
-                return func(namespace_name, repo_name, *args, **kwargs)
-
-            repository = namespace_name + "/" + repo_name
-            if allow_public:
-                repository_ref = registry_model.lookup_repository(namespace_name, repo_name)
-                if repository_ref is None or not repository_ref.is_public:
-                    raise Unauthorized(repository=repository, scopes=scopes)
-
-                if repository_ref.kind != "image":
-                    msg = (
-                        "This repository is for managing %s and not container images."
-                        % repository_ref.kind
-                    )
-                    raise Unsupported(detail=msg)
-
-                if repository_ref.is_public:
-                    if not features.ANONYMOUS_ACCESS:
-                        raise Unauthorized(repository=repository, scopes=scopes)
+    def _require_permission(allow_for_superuser=False, disallow_for_restricted_users=False):
+        def wrapper(func):
+            @wraps(func)
+            def wrapped(namespace_name, repo_name, *args, **kwargs):
+                logger.debug(
+                    "Checking permission %s for repo: %s/%s",
+                    permission_class,
+                    namespace_name,
+                    repo_name,
+                )
+
+                # In the cases where we need to act based on the user to workaround the regular permissions
+                # and not just based on the validity of the JWT, we need to hit the database to get the user.
+                # This need to be checked before the normal permissions, in order to restrict a user's own namespace
+                if features.RESTRICTED_USERS and disallow_for_restricted_users:
+                    if allow_public:
+                        repository_ref = registry_model.lookup_repository(namespace_name, repo_name)
+                        if repository_ref is None or not repository_ref.is_public:
+                            raise Unauthorized()
+
+                        if not features.ANONYMOUS_ACCESS:
+                            raise Unauthorized()
+
+                        return func(namespace_name, repo_name, *args, **kwargs)
+
+                    context = get_authenticated_context()
```

Comment:
NIT: If we define context outside the `if` block, we don't have to define it again in the superuser if condition block. Also, if context is `None`, how do we handle it?

Reply:
I didn't putting it outside the if block in order to avoid making a database call for the user if the features were not enabled. If the context is None, then the regular anonymous flow should still apply.